Soft k-means example in 1D (#9)
tweag/chainsail-resources: examples/soft-kmeans/README.md

# Soft k-means clustering

> Soft k-means, also known as fuzzy clustering, is the probabilistic version of the better knwon k-means clustering.

Author: ```suggestion
[Soft k-means clustering](https://en.wikipedia.org/wiki/Fuzzy_clustering), also known as fuzzy clustering, is the probabilistic version of the better known k-means clustering.
```

In this example, we simulate a one dimension clustering problem to illustrate a use-case of Chainsail.

## Data generation
We generated two 1-dimension clusters of points by drawing them from two normal probability distributions:
- **Cluster 1:** 80 points, drawn for a normal probability distribution of mean `mu1 = 0` and standard deviation `sigma = 1`
- **Cluster 2:** 120 points, drawn for a normal probability distribution of mean `mu2 = 1.5` and standard deviation `sigma = 1`

<img alt="data points" src="./images/data_points_1D.png" height="400"/>

## Model
The model used to fit those clusters is a gaussian mixture (inspired from the [Stan example of soft k-means](https://mc-stan.org/docs/2_21/stan-users-guide/soft-k-means.html)).
It is defined in the file `probability.py`.  

> - **Number of mixture components:** Just like in regular "hard" k-means, the users defines the number of clusters to find a priori. There are two clusters to find in this case, so the mixture model was defined with 2 components.

Reviewer: Can we label the clusters in the plots? Like, with texts saying "cluster 1" and "cluster 2"?

Author: There are labelled with a legend on the first plot. Then do you think we should show the underlying true clusters on the subsequent plots as well ?

- **Components weights:** Since the clusters do not contain the same number of points, weights were set on the components of the mixture at `w1 = 0.4` and `w2 = 0.6` (which fits the proportions of data points in each cluster).

> - **Parameters to fit:** To keep things simple, only the means of the components were set a variable parameters. The standard deviations were fixed at `sigma = 1` (which fits the data generation model), and components weights were also kept fixed.

Author: I think it's nice to keep it explicit, for the unexperienced clusterer (I did not know this was standard procedure myself for example).

As for the standard deviation, are you sure that there never is a reason to fit it as well? If a cluster was generated with a sd that is very different from the one fixed in the model (not the case here), I would expect it to produce biased results.

- **Prior parameters:** The prior parameters for both components of the gaussian mixture were arbitrarily set at mean `mu=0` and standard deviation `sigma = 5`.

The log-posterior distribution is visualized below:

> <img alt="posterior" src="./images/posterior_1D.png" height="400"/>

Author: You're right, that's a leftover from previous version.

> The posterior shows two modes, which correspond to the two possible positions of the mixture model to explain the generation of the clusters (Situation 1: components 1 and 2 of the gaussian mixture generated cluster 1 and 2 respectively. Situation 2 (reverse): components 1 and 2 generated clusters 2 and 1 respectively).

Reviewer: That's a very long sentence. What about
```suggestion
The posterior shows two modes, each of which correspond to one possible parameter configuration explaining the data: in the first configuration, components 1 and 2 of the mixture generate cluster 1 and 2 respectively.
In the second configuration, the roles of the mixture components are reversed: component 1 generates cluster 2 and component 2 generates cluster 1.```

> But since the clusters contain a different number of points, and the mixture components have different weights, one situation is less likely to be true than the other.

Reviewer: ```suggestion
But since the clusters contain a different number of data points and the mixture components have different weights, one configuration is less likely to be correct than the other.
```

This is highlighted by the fact that the top-left mode in the posterior shows a higher probability than the lower-right mode.

## Single MCMC chain sampling

> We created a simple single MCMC chain for sampling this posterior distribution, defined in the module `single_chain.py`.

Reviewer: Let's highlight here that this is essentially what Stan and other PPLs do, although they do give the user the possibility to run several (isolated) chains.

### Sampling

> With an initial state for the MCMC chain set at (1, 0), the chain gets stuck on the lower probability mode.

Reviewer: ```suggestion
With an initial state for the MCMC chain set at (1, 0), the chain gets stuck in the lower probability mode.
```

> That means that the gaussian mixture component with the higher weight (component 2) fits the cluster with the lower number of points, and vice versa.

Reviewer: ```suggestion
That means that the mixture component with the higher weight (component 2) fits the cluster with the lower number of points, and vice versa.
```

<img alt="sampling single chain with initial state = (1, 0)" src="./images/sampling_sc_initstate-1-0.png" height="400"/>

> This results in the samples of the means being slightly offset from their expected positions (0 and 1.5), as to compensate for the mismatch.

Reviewer: We should add a sentence that tells the reader that this is what (and what else, namely a histogram of the sampled means) the next plot shows.

<img alt="sampled means single chain" src="./images/samples_means_sc_initstate-1-0.png" height="400"/>

### Cluster assignment
Soft k-means clustering treats the cluster assignments as probability distributions over the clusters.
Each data point is associated with a log unnormalized probability distribution for each cluster, which is obtained from transformation of the samples of the means (called the "soft Z" transformed samples).

> These soft Z distributions, after softmax normalization, provide the probability of each point to belong to each cluster.

Reviewer: We could call this probability the `responsibility` that a component takes for explaining a data point, in line with the excellent discussion in Bishop's "Pattern Recognition and Machine Learning", p. 432, and even point the reader to that reference.

The distributions of the soft Z samples for 3 individual data points are shown below (one point located on the left, one in the middle, and one on the right).

> <img alt="Soft_z analysis" src="./images/soft-z_samples_sc2.png"/>

Reviewer: Can we remove the redundant "Point" in the subplot titles? Or, if we keep it, name the points, like Point 1, Point 2, Point 3 in the subplot titles. Or even just write x_!, x_2, x_3 :thinking:

> The leftward point at `x = -0.687`, shows a distribution of higher soft Z values for the cluster 2, that is clearly distinct for the distribution of soft Z values for cluster 1.

Reviewer: ```suggestion
The leftward point at `x = -0.687` shows a distribution of higher soft Z values for cluster 2 and which is clearly distinct from the distribution of soft Z values for cluster 1.
```

Therefore, it can be unequivocally assigned to the cluster 2.
Similarly, the rightward point at `x = 2.369` can be unequivocally assigned to cluster 1.

> However, the middle point at `x = 1.4` displays two distributions of soft Z values that are superimposed.

Reviewer: ```suggestion
However, for the middle point at `x = 1.4`, the two soft Z-distributions overlap strongly.
```

> This highlights the fact that this point is in an ambiguous area, and cannot be assigned confidently to any of the two cluster.

Reviewer: ```suggestion
This highlights the fact that this point is in an ambiguous area and cannot be assigned confidently to either cluster.
```

> Note that this uncertainty is not quantified in the traditional "hard" k-means algorithm, whereas it is available here, through these soft Z samples.

Reviewer: ```suggestion
Note that this uncertainty is not quantified in the traditional "hard" k-means algorithm.
```

> This being said, the soft Z distributions can be summarized to their median value, and consequently, the soft k-means clustering can be converted to a "hard" k-means by assigning each point to its cluster of higher median probability.

Reviewer: ```suggestion
If we summarize the soft-Z distributions by their median value, soft k-means clustering can be converted to a "hard" k-means by assigning each point to its cluster of highest median probability.
```

<img alt="cluster assignment single chain" src="./images/cluster_assignment_sc_initstate-1-0.png" height="400"/>

In this case, because the lower probability mode of the posterior is sampled, the cluster assignment is rather imperfect.
The points at the frontier of the 2 clusters (framed in dotted line) were assigned to the left cluster, which is the less probable one.

## Chainsail sampling

> 

Reviewer: Let's summarize here in once sentence what went wrong above and why we expect Chainsail to do better.

### Sampling

> We also sampled the posterior distribution with Chainsail.
> This time, replica exchange allows the main MCMC chain to not get stuck in one mode, and to sample both the low and the high probability modes.

Reviewer: ```suggestion
We now sample the posterior distribution with Chainsail.
This time, Chainsail's Replica Exchange algorithm allows the main MCMC chain to sample both the low and the high probability mode instead of getting stuck in one of them.
```

> <img alt="sampling chainsail with initial state = (1, 0)" src="./images/sampling_chainsail_initstate-1-0.png" height="400"/>

Reviewer: Perhaps we can show only a subset of the samples, such that the color of the modes is still a bit visible.

Author: I re-overlayed the posterior on top of it with transparency to better show the colors, which is the best I could get imo.

Author: I re-overlayed the posterior on top of it with transparency to better show the colors, which is the best I could get imo.

The sampled means are shown below.

> The two higher peaks correspond to the samples from the higher posterior mode, and the lower peaks to samples from the lower one.

Reviewer: ```suggestion
The two higher peaks correspond to the samples from the higher posterior probability mode, and the lower peaks to samples from the lower one.
```

> Similarly to the samples from the single chain, the means sampled from the lower posterior mode are offset from their expected position, but more importantly, samples from the higher posterior mode match the theoretical cluster centroids much better.

Reviewer: We should tell the reader why the cluster centroids are matched better.

<img alt="sampled means chainsail" src="./images/samples_means_chainsail_initstate-1-0.png" height="400"/>

> 

Reviewer: We can stress again here how different this sampling result is from the one obtained with a single chain.

Author: I highlighted the difference once more in another part (right below), so I'll leave this one like that. I think it is clear enough already.

Author: I highlighted the difference once more in another part (right below), so I'll leave this one like that. I think it is clear enough already.

### Cluster assignment
Similarly, the soft Z samples show bimodal distributions.

> <img alt="Soft_z analysis" src="./images/soft-z_samples_chainsail2.png"/>

Reviewer: Maybe it's just too late in the evening, but I currently have the feeling that this plot is the most convincing result here! Essentially, it shows that, with Chainsail, we capture the fact that there is a certain probability for _both_ component 1 and 2 to have created, e.g., the left-most point. This is absent in the soft-Z histograms for the single chain, and thus sampling with a single chain made us miss out on important information :slightly_smiling_face: If you agree, let's highlight this important result clearly to the reader here.

And finally, the clusters assignments are closer to what would be expected with a hard k-means clustering.

> <img alt="cluster assignment chainsail" src="./images/cluster_assignment_chainsail_initstate-1-0.png" height="400"/>

Reviewer: This is also a very nice and consistent result and a great finish to this post - back to the roots of (correct) hard k-means.

> 

Reviewer: We definitely need a conclusion-like paragraph here in which we summarize these results and maybe link out to additional resources such as the Chainsail website (+ a future announcement blog post), possibly the Replica Exchange blog post, repeat the reference to the Bishop book, etc.
And motivate the reader to think about more multimodal problems and try Chainsail on them :slightly_smiling_face:

## Reproduce the example

> - The code to generate those figures lives in the jupyter notebook `soft-kmeans.ipynb`. It is packaged with [Nix](https://github.com/NixOS/nix) through [JupyterWith](https://github.com/tweag/jupyterWith). It can be run with the following command:

Reviewer: ```suggestion
- The code to generate those figures lives in the Jupyter notebook `soft-kmeans.ipynb`. It is packaged with [Nix](https://github.com/NixOS/nix) through [JupyterWith](https://github.com/tweag/jupyterWith). It can be run with the following command:
```

Reviewer: Let's add an (absolute) link to the example directory in this repository; after all chances are that this post will be published on the Tweag website, which removes the context of the `chainsail-resources` repository (although the post + notebook can still live in there).

Note that single chain sampling is also what most probabilistic programming libraries provide by default, although in a more advanced manner: they usually give the possibility to mix samples from multiple isolated chains and provide some sort of initial state estimation, which in practice makes these algorithms a bit more robust than our naive implementation here.

### Sampling
With an initial state for the MCMC chain set at `(mean of 1st component, mean of 2nd component) = (1, 0)`, the chain gets stuck in the lower probability mode:

<img alt="sampling single chain with initial state = (1, 0)" src="./images/sampling_sc_initstate-1-0.png" height="400"/>

That means that the mixture component with the higher weight (component 2) incorrectly fits the cluster with the lower number of points and vice versa.
As the following histogram shows, this results in the samples of the means being slightly offset from their expected positions (`~(0, 1.5)`) due to the mismatch between the number of points in the cluster and the weight of the Gaussian mixture component:

> <img alt="sampled means single chain" src="./images/samples_means_sc_initstate-1-0.png" height="400"/>

Author: No strong opinion. I generally prefer to have all axis properly labeled, but I'm ok with removing those here.

This demonstrates that Chainsail found _both_ modes - the "correct", desired one _and_ the less likely one and thus correctly and honestly samples the full posterior distribution.

### Cluster assignment
Similarly, the soft Z samples show bimodal distributions:

<img alt="Soft_z analysis" src="./images/soft-z_samples_chainsail2.png"/>

Therefore, with Chainsail, we capture the fact that there is a certain responsibility for *both* component 1 and 2 to have generated each data point.
This is absent in the soft Z histograms for the single chain, and thus single-chain sampling made us miss out on important information. 
Finally, the clusters assignments are much closer to what would be expected with a hard k-means clustering:

<img alt="cluster assignment chainsail" src="./images/cluster_assignment_chainsail_initstate-1-0.png" height="400"/>

## Conclusion
While this example showcased a rather trivial, one-dimensional problem, it highlighted the difficulty for single-chain MCMC algorithms to correctly sample multimodal posteriors, and the biased analysis results that this leads to.

> Keep in mind that while the posterior distribution is easily to visualize in this 1D case, most real life problems generate highly dimensional posteriors, in which case a bias in sampling is much less obvious to detect.

Author: ```suggestion
Keep in mind that while the posterior distribution is easy to visualize in this 1D case, most real life problems generate highly dimensional posteriors, in which case a bias in sampling is much less obvious to detect.
```

> # Soft k-means clustering

Author: Sure, your suggestion sounds good to me.
```suggestion
# Better soft k-means clustering with Chainsail - multimodality treated right
```
